agents/ModularAgent.py

```diff
@@ -0,0 +1,649 @@
+from pprint import pprint
+from random import random
+from random import choice
+
+from concept_formation.preprocessor import Flattener
+from concept_formation.preprocessor import Tuplizer
+from concept_formation.structure_mapper import rename_flat
+
+
+from agents.BaseAgent import BaseAgent
+from learners.WhenLearner import get_when_learner
+from learners.WhereLearner import get_where_learner
+from learners.WhichLearner import get_which_learner
+from planners.fo_planner import FoPlanner, execute_functions, unify, subst
+# Chris Questions:
+# Why does the explanation have 'value' deference instead of there being a whole path to the value in the resolved thing?
```

Comment:
Have these questions been sufficiently answered?

Reply:
**Why does the explanation have 'value' deference [sic] (dereference) instead of there being a whole path to the value in the resolved thing?**
I think this question would be more concretely stated as why is every how part something like ('value', ( <actual how expression>)). I think this just seemed odd to me because this LISP-like tuple language we've got going on has no syntactic differences between an assignment, a dereference, and an application of an operator. And in any given RHS you've got all three at once. I suppose this choice has a lot to do with the fo_planner, but it certainly wasn't obvious when I first looked at it.

 **How can we make where an actual thing?**
I don't remember exactly what I was thinking with this, but I think the question had to do with the state of existing where learners and their ability to create generalizations. I think this question would be best answered by trying to do some concrete things and thinking about what formulation of the where learner would solve any issues that arise efficiently enough to be practical. 

**Calling all inputs and outputs foas is a little confusing?**
I have not totally addressed this. In code the word 'foa' is only used for the user given foas. The container which keeps track of the internal variables for each RHS is now called input_vars, but the string values for those are still something like '?foa#'. Perhaps '?arg#" would better convey this. "?foa0" used to always refer to the selection name, this is now just '?selection' for the sake of clarity. 

**In training why does HowSearch run at the beginning and then only get used when there are no explanations?**
I have fixed this, although the way that the RHS is executed is a little tricky (I suppose this is why Chris did it in this way originally -- as a quick and dirty fix).  Basically, my half fix for this was to just loop over all the operators until one matched the expression in the RHS. A more efficient way to do this would be to have that expression include an executable object. I have a pytorch version of the how learner which could be a basis for this, although perhaps we should have some general wrapper object so that we don't impose any requirements on the how learner.   

**Secondary explanations... what to do....**
Ok so this is weird and @cmaclell will probably need to chime in on this. So the way things work right now is that in explanations_from_skills() the where learner produces some matches and then the when learner checks them. Then later where_matches() is called and does where_learner.check_match() to separate the explanations from the nonmatching_explanations (used to be 'secondary_explanations' which was a superset of explanations). Note that before all of this happened in one loop, and functionally it is still kind of in one loop because explanations_from_skills() returns an iterator.  However, the weird thing is that there are never any nonmatching_explanations. I literally put in an error statement to break the whole thing if any were ever created. So my question for Chris would be "in what cases could get_matches() ever produce a match that failed a subsequent call to check_match()?" I have been using the MostSpecific _where_ all this time so let me know if there is some odd case where this is possible. If it is not possible we should remove this whole part about secondary/nonmatching explanations

Reply:
Chiming in on the secondary explanations.

So this might happen when explaining a users demonstration.

For example, imagine we have a skill S1 for adding two numbers and putting the result in a third field and we see an example E1 which is also the sum of two numbers, but where the the numbers are coming from two fields we've never observed numbers coming from before.

In this case, the where and when learners will not match (specifically the most specific where). However, if we just took the right hand side (the summing of two numbers), we can find that it can explain the example E1. This unrestricted check of the RHS against the example E1 is where the secondary explanations come from. 

It is better to just try using the right hand side without any restrictions from the where and when (e.g., basically just using the legality conditions of the RHS) than doing a general how search with all the operators. Specifically, it is much faster to search to a depth of 1 over previously used sequences than it is to do a how search to depth >1 over all the overly general operators.

Reply:
@DannyWeitekamp to tie this to your specific question. Secondary explanations or... explanations_from_skill?... should not be checking where and when conditions, it should only check if the RHS of the skill (?? really confused about our terminology now, so not sure you're calling the combination of where, when and how outputs a skill anymore) matches.

Make sense?

Reply:
What was secondary_explanations is now nonmatching_explanations. I thought this was a clearer name given the structure of the original code [here](https://github.com/cmaclell/apprentice_learner_api/blob/master/agents/WhereWhenHowNoFoa.py#L278)
I'm seeing now that I may have made a mistake in terms of replicating things. This may account for the 5x speedup in my version. My explain_sai() does actually use the where_learner, but not the when whereas yours uses neither. Your version was actually doing this depth >1 search, but if I'm hearing you correctly, what you intended for it to do (eventually) was to start by doing a depth=1 search over the existing RHSs and then do a full search at the specified depth over all of the operators. I think this is motivation enough for making some sort of executable object that encompasses both operators and chains of them. I'll find some way of regaining this functionality, I should also use this as an opportunity to modularize the _how_.